### ISSUE TITLE
feature request - shift to lazily loaded resource relationships in some way?

### ISSUE THREAD

**Issue by acorncom** [tuyakhov/yii2-json-api#34]
Version |
-------- |
1.0.3

TL;DR: In scenarios where there are many relationships but a client only wants to refresh a single relationship, the serializer seems to require eager-loading all relationships before filtering down to a specific relationship to include.

In the theoretical example below:

```php
class Blog implements ResourceInterface, LinksInterface
{
    use ResourceTrait;

    public function extraFields()
    {
        return [
            'author',
            'comments',
            'likes',
        ];
    }

    public function getAuthor()
    {
        return $this->hasOne(Author::className(), ['id' => 'author_id']);
    }

    public function getComments()
    {
        return $this->hasMany(Comment::className(), ['id' => 'blog_id']);
    }

    public function getLikes()
    {
        // complicated SQL packaged up into a model runs here ...
    }
}
```

If an API client wants to do a `/api/blogs/3?include=comments` without triggering the complicated SQL in the `likes` relationship, the `getLikes()` will still have been loaded using the default `ResourceTrait` and default `Serializer` due to the `resolveFields()` `foreach` here ...

https://github.com/tuyakhov/yii2-json-api/blob/f2bee3dd7201d1e2bfe50a666582e53d401f7d0f/src/ResourceTrait.php#L56-L68

Specifically, line 65 there eager-loads all relationships in the `extraFields` method even if those relationships are not included in any way ...

Do you think we can work out how to not pre-populate those relationships until we're sure we want them in the payload? Combining that desire with the option to add links could get a bit tricky ...

Interested in your thoughts!

**Comment by tuyakhov** [comment]
Hi David,
Thank you for reporting this 
Yes, originally it was designed like this in order to add links to the payload as it is described in the specification:
```
A “relationship object” MUST contain at least one of the following:
 - links
 - data
 - meta
```
But I went through the specification and also checked the code once again. And I discovered that the `links` object doesn't require the relationship itself to be populated. You can check that here https://github.com/tuyakhov/yii2-json-api/blob/master/src/ResourceTrait.php#L94
Please correct me if I wrong, but I think that avoiding pre-population is something that could be implemented easily. It gets a bit tricky if we want to support dot-separated path, like `?include=comments.author`

**Comment by acorncom** [comment]
Hi Anton,
Yes, I think you're looking at it right. In fact, I think we should be able to support either related links (as we're doing) or allow users to generate direct links (i.e. `/api/authors/1`) without loading the relationship (as it's possible to use the `$blog->author_id` instead of `$blog->author->primaryKey`)

But yes, I think we do want to support `?include=comments.author` relationships as well (I hit a need for that yesterday 😀 ). For that's where the power of the includes system really comes into play ...
